explainer_comparison/explainer_utilities.py

```diff
@@ -24,3 +24,36 @@ def run_and_collect_explanations(factory: ExplainerFactory, X_data, explainers=N
         return pd.concat(results, axis=1)
     else:
         return pd.DataFrame()  # Return an empty DataFrame if no explanations were added
+
+
+def permutation_feature_importance(model, X_data, y_data, metric='accuracy', random_state=None):
+    """
+    Calculates permutation feature importance for a given model.
+    
+    Parameters:
+    - model: The trained machine learning model.
+    - X: The feature matrix.
+    - y: The target vector.
+    - metric: The metric to use for evaluating the model. Either 'accuracy' or 'mse'.
+    - random_state: The random seed for reproducibility.
+    
+    Returns:
+    - feature_importances_df: A DataFrame containing the feature names and their importance scores, sorted by scores.
+    """
+    if metric == 'accuracy':
+        baseline_score = accuracy_score(y_data, model.predict(X_data))
+        scorer = accuracy_score
+    elif metric == 'mse':
+        baseline_score = mean_squared_error(y_data, model.predict(X_data))
+        scorer = mean_squared_error
+    else:
+        raise ValueError("Invalid metric. Please choose 'accuracy' or 'mse'.")
+    
+    feature_importances = {}
+    for feature in X_data.columns:
+        X_data_permuted = X_data.copy()
+        X_data_permuted[feature] = np.random.permutation(X_data[feature])
+        permuted_score = scorer(y_data, model.predict(X_data_permuted))
+        feature_importances[feature] = baseline_score - permuted_score
```

Comment:
the importance by the permutation importance method goes as follows:
the more the intervened model decreases in its performance compared to the original model - the permutated feature is more important.

however, the explanations based on this method are as follows:
the local effect of each feature is measured by taking the differencing between the original local prediction to the intervened model local prediction:
original model predict(x) outcome minus intervened model predict(x).

so please provide API also to get the explanations

Reply:
<img width="745" alt="Screenshot 2024-06-10 at 12 43 41" src="https://github.com/emunaran/explainer-comparison/assets/106267539/c89fff74-c885-4933-9326-f717a361b449">

Ran, to confirm my understanding: instead of using a specific error measure as described in [the book](https://christophm.github.io/interpretable-ml-book/feature-importance.html), we are to compare the predictions directly by computing the difference between the original and the permuted predictions for each feature, correct?

If that's the case, would it be appropriate to aggregate these differences using the mean, such that the feature importance calculation would look like this:

original_predictions = model.predict(X_data)
permuted_predictions = model.predict(X_data_permuted)
feature_importances[feature] = np.mean(original_predictions - permuted_predictions)

?

Reply:
we need both:
a. measure the importance as you did (by the book)
b. measure the feature impact, yes, as you described above.
the function should then return both a and b.